Report errors to sentry

fvictorio opened the pull request.

alcuadrado commented on packages/buidler-core/src/internal/cli/sentry/subprocess.ts:
```diff
@@ -0,0 +1,96 @@
+  try {
+    event = JSON.parse(serializedEvent);
+  } catch (error) {
+    log("BUIDLER_SENTRY_EVENT env variable doesn't have a valid JSON, exiting");
```
I'd log the `serializedEvent` here. That'd help us debug this remotely if it ever happens

fvictorio replied:
Done.

alcuadrado commented on packages/buidler-core/src/internal/cli/sentry/subprocess.ts:
```diff
@@ -0,0 +1,96 @@
+
+  anonymizePaths(event, configPath);
+
+  Sentry.captureEvent(event);
```
I thought `Sentry.close` was needed after a `capture*`. Doesn't it?

alcuadrado replied:
Also, I guess any `Sentry` method can throw, so I'd wrap all of those in a try/catch and log any error. wdyt?

fvictorio replied:
> I thought Sentry.close was needed after a capture*. Doesn't it?

I think here it's not necessary. In the main process we have to do it because there's an explicit call to `process.exit`.

> Also, I guess any Sentry method can throw, so I'd wrap all of those in a try/catch and log any error. wdyt?

Yes, will do that.

alcuadrado replied:
Oh, that's cause the standard integrations aren't disabled here, right? Won't that capture the `log()` calls and also report them?

fvictorio replied:
Done.

alcuadrado commented on packages/buidler-core/src/internal/cli/sentry/transport.ts:
```diff
@@ -0,0 +1,37 @@
+        env.BUIDLER_SENTRY_CONFIG_PATH = configPath;
+      }
+
+      fork(`${__dirname}/subprocess`, [], {
```
I think this has to be detached, at least in Linux and macos. Here's more info about it: https://nodejs.org/api/child_process.html#child_process_options_detached

I'd also test what `detached` does on windows.

fvictorio replied:
Done.

fvictorio commented:
I fixed the comments and converted the Reporter to a singleton. Now I'll test this on windows, but in the meantime I think the latest commits can be reviewed.

fvictorio commented on packages/buidler-core/src/internal/cli/cli.ts:
```diff
@@ -181,6 +185,15 @@ async function main() {
 
+    const reporter = Reporter.getInstance();
+    try {
+      if (config === undefined || config.analytics.enabled) {
```
This seems weird, but the point is that if an error was thrown before the config could be loaded, we err on the side of reporting it.

alcuadrado replied:
I see. This makes lots of sense. I would just add a comment explaining it, so we don't accidentally delete it in the future.

fvictorio commented on packages/buidler-core/src/internal/cli/cli.ts:
```diff
@@ -44,6 +45,8 @@ async function main() {
   // stack traces before really parsing the arguments.
   let showStackTraces = process.argv.includes("--show-stack-traces");
+  let verbose = false;
+  let config: ResolvedBuidlerConfig | undefined;
```
Previously I had "hoisted" the config path. I could've done the same with the `analytics.enabled` value but... it was getting annoying. I don't know if there's a better way to do this (share a value between the try and catch blocks).

alcuadrado commented on packages/buidler-core/src/internal/cli/sentry/anonymizer.ts:
```diff
@@ -0,0 +1,116 @@
+    const result: StackFrame = {
+      lineno: frame.lineno,
+      colno: frame.colno,
+      function: frame.function,
```
I think keeping these attributes for non-anonymous/Buidler frames would be very useful for debugging different issues in different versions:

`context_line`
`pre_context`
`post_context`
`vars`

More info here: https://develop.sentry.dev/sdk/event-payloads/stacktrace/

alcuadrado commented:
There are a few `log()` calls with errors in these files, which I think should also be reported:
* https://github.com/nomiclabs/buidler/blob/development/packages/buidler-core/src/builtin-tasks/node.ts
* https://github.com/nomiclabs/buidler/blob/development/packages/buidler-core/src/internal/buidler-evm/provider/node.ts
* https://github.com/nomiclabs/buidler/blob/development/packages/buidler-core/src/builtin-tasks/utils/watch.ts

fvictorio commented on packages/buidler-core/src/internal/sentry/anonymizer.ts:
```diff
@@ -0,0 +1,152 @@
+  private _anonymizeFilename(
+    filename: string
+  ): { anonymizedFilename: string; anonymizeContent: boolean } {
+    if (filename === this._configPath) {
```
This function is very complex and could use some tests.

If `packageJsonPath` is made a parameter instead of being computed, then this function becomes pure (or, well, pure enough) and testing it would be very straightforward. The downside is that it makes things weird for the callee (it has to call `findup` for all files) and we have to make this method public (which I don't mind).
